### PR TITLE
refractor(core/compiler): hide internal types from generated code

### LINKED ISSUE #40676: Do not expose internal types in published APIs (mark them as `unknown`).

### PR DESCRIPTION
**Pull request by petebacondarwin**
Fixes #40676

*This PR is marked as `major` because the names of typings that appear in libraries compiled by ngcc have changed. This is not a breaking change for general developing (building and testing applications and libraries) but could affect 3rd party tooling who might have been relying upon these types generated by ngcc. By putting this commit in a major release (i.e. the forthcoming 12.0.0) it gives them time and warning to update the tooling to handle the change if necessary.*

---

~~This PR sits on top of #41080 so you only need to review from **docs(core): fix typos in interface docs** onwards.~~

### CODE REVIEW

**Review by mhevery — APPROVED** [review]

File: packages/core/src/render3/interfaces/public_definitions.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+/**
+ * @publicApi
+ */
+export type ɵɵFactoryDeclaration<T, CtorDependencies extends CtorDependency[]> = unknown;
+
+/**
+ * An object literal of this type is used to represent the metadata of a constructor dependency.
+ * The type itself is never referred to from generated code.
+ *
+ * @publicApi
```
Comment (JoostK) [inline packages/core/src/render3/interfaces/public_definitions.ts:56]: This one is interesting. The `CtorDependency` type isn't exported, nor should it be. Maybe `@codeGenApi` is more appropriate?

We should at some point include all types that govern the partial compilation output in golden files, in which case this type should be included, but the type itself shouldn't be publicly exported.

**Review by JoostK — APPROVED** [review]

File: packages/core/src/render3/interfaces/public_definitions.ts (hunk 1 above)
Comment (petebacondarwin) [reply packages/core/src/render3/interfaces/public_definitions.ts:56]: So I would leave the `@publicApi` jsdoc tag on it. This tag really only indicates that it should appear in a golden file.
(The doc-gen doesn't care about it AFAIK.)

Comment (JoostK) [reply packages/core/src/render3/interfaces/public_definitions.ts:56]: I believe only exported types end up in the golden file, the `@publicApi` marker alone is not sufficient (notice how this PR doesn't update the goldens for `@angular/core`, whereas `CtorDependency` wasn't previously public)

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:363496537:BASE:363496573:1616013422408:29fc17c9).

File: packages/core/src/render3/interfaces/public_definitions.ts (hunk 1 above)
Comment (petebacondarwin) [reply packages/core/src/render3/interfaces/public_definitions.ts:56]: Yes, two hurdles must be overcome I believe: it must be exported from the barrel and marked with `@publicApi`.

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:363496537:BASE:363935484:1616178980019:aa9c7cb4).

**Pull request closed without merging**